File: 프로그래머스/2/64065. 튜플/튜플.java
import java.util.*;

class Solution {
    public int[] solution(String s) {
        int[] answer = {};
        List<List<String>> tuples = new ArrayList<>();
        
        char[] chars = s.toCharArray();
        List<String> tuple = new ArrayList<>();
        char pre = 'z';
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < chars.length; i++) {
            if (i == 0 || i == chars.length - 1) {
                continue;
            }
            
            if (chars[i] == '{') {
                tuple = new ArrayList<>();
                sb = new StringBuilder();
            }
            
            if (chars[i] == '}') {
                tuple.add(sb.toString());
                tuples.add(tuple);
            }
            
            if (chars[i] == ',') {
                if (pre == '}') {
                    continue;
                } else {
                    tuple.add(sb.toString());
                    sb = new StringBuilder();
                }
            } 
            
            if ('0' <=  chars[i] && chars[i] <= '9') {
                sb.append(String.valueOf(chars[i]));
            }
            pre = chars[i];
        }
        Collections.sort(tuples, (i1, i2) -> i1.size() - i2.size());
        
        answer = new int[tuples.size()];
        Set<Integer> check = new HashSet<>();
        for (int i = 0; i < tuples.size(); i++) {
            List<String> t = tuples.get(i);
            for (String s2 : t) {
                int n = Integer.parseInt(s2);
                if (!check.contains(n)) {
                    answer[i] = n;
                    check.add(n);
                    break;
                }
            }
        }
        return answer;
    }
}

File: 프로그래머스/2/64065. 튜플/튜플.py
# 튜플의 표현을 통해 튜플을 추축하는 문제

# 튜플의 길이로 정렬
# 튜플의 반복문을 돌면서 answer에 없는 원소를 넣기
def solution(s):
    answer = []
    tuples = changeTuple(s)
    tuples = sorted(tuples, key = lambda x: (len(x)))
    
    for tp in tuples:
        for number in tp:
            number = int(number)
            if number not in answer:
                answer.append(number)
    
    return answer

def changeTuple(s):
    s = s[1:-1]
    index = 0
    tuples = []
    while index < len(s):
        n = s[index]
        if n == "{":
            check = []
            number = ""
            while n != "}":
                if n.isdigit():
                    number += n
                elif n == ",":
                    check.append(number)
                    number = ""
                index += 1
                n = s[index]
            check.append(number)
            tuples.append(check)
        else:
            index += 1
    
    return tuples